pkg/agent/proxy/proxier.go

```diff
@@ -155,17 +156,26 @@ func (p *proxier) removeStaleServices() {
 				continue
 			}
 		}
-		// Remove Service group whose Endpoints are local.
-		if svcInfo.NodeLocalExternal() {
+		// If externalTrafficPolicy of the Service is Local and internalTrafficPolicy of the Service is not Local, a Service group
+		// that only has local Endpoints should have been already created, and it should be removed.
+		if svcInfo.NodeLocalExternal() && !(p.svcInternalTrafficPolicyEnabled && svcInfo.NodeLocalInternal()) {
 			groupIDLocal, _ := p.groupCounter.Get(svcPortName, true)
 			if err := p.ofClient.UninstallServiceGroup(groupIDLocal); err != nil {
 				klog.ErrorS(err, "Failed to remove flows of Service", "Service", svcPortName)
 				continue
 			}
 			p.groupCounter.Recycle(svcPortName, true)
 		}
-		// Remove Service group which has all Endpoints.
-		groupID, _ := p.groupCounter.Get(svcPortName, false)
+
+		var groupID binding.GroupIDType
+		// If internalTrafficPolicy of the Service is Local, a Service group that only has local Endpoints should have been
+		// already created. If internalTrafficPolicy of the theService is Cluster, a Service group that has all Endpoints
+		// should have been created. Remove the Service group.
+		if p.svcInternalTrafficPolicyEnabled && svcInfo.NodeLocalInternal() {
+			groupID, _ = p.groupCounter.Get(svcPortName, true)
+		} else {
+			groupID, _ = p.groupCounter.Get(svcPortName, false)
+		}
```

Comment:
what about the call to `p.groupCounter.Recycle` below? I only see a call to `p.groupCounter.Recycle(svcPortName, false)`, is that normal?

I feel like the code could benefit from a small refactoring for improved readability. Maybe something along these lines:
```go
hasServiceGroupWithLocalEndpoints := svcInfo.NodeLocalExternal() || (p.svcInternalTrafficPolicyEnabled && svcInfo.NodeLocalInternal())
hasServiceGroupWithAllEndpoints := !svcInfo.NodeLocalExternal() || !(p.svcInternalTrafficPolicyEnabled && svcInfo.NodeLocalInternal())

if hasServiceGroupWithLocalEndpoints {
    // ...
}

if hasServiceGroupWithAllEndpoints {
    // ...
}
```